api/plgd_time.c

```diff
@@ -0,0 +1,415 @@
+{
+  const oc_string_t *lst_rfc3339 = NULL;
+  for (; rep != NULL; rep = rep->next) {
+    if (rep->type == OC_REP_STRING &&
+        oc_rep_is_property(rep, PLGD_TIME_PROP_LAST_SYNCED_TIME,
+                           OC_STRLEN(PLGD_TIME_PROP_LAST_SYNCED_TIME))) {
+      lst_rfc3339 = &rep->value.string;
+      continue;
+    }
+    OC_WRN("plgd time: unknown property (%s:%d)", oc_string(rep->name),
+           (int)rep->type);
+  }
+
+  if (lst_rfc3339 == NULL) {
+    OC_ERR("cannot decode plgd time: property %s not found",
+           PLGD_TIME_PROP_LAST_SYNCED_TIME);
+    return false;
+  }
+
+  oc_clock_time_t lst = oc_clock_parse_time_rfc3339(
+    oc_string(*lst_rfc3339), oc_string_len(*lst_rfc3339));
+  if (lst == 0) {
+    return false;
+  }
+  pt->store.last_synced_time = lst;
+  return true;
+}
+
+static int
+store_encode_plgd_time(size_t device, void *data)
+{
+  (void)device;
+  const plgd_time_t *pt = (plgd_time_t *)data;
+  return plgd_time_encode(*pt, OC_IF_RW, /* to_storage */ true);
+}
+
+bool
+plgd_time_dump(void)
+{
+  long ret = oc_storage_save_resource(PLGD_TIME_STORE_NAME, /*device*/ 0,
+                                      store_encode_plgd_time, &g_oc_plgd_time);
+  if (ret <= 0) {
+    OC_ERR("cannot dump plgd time to storage: error(%ld)", ret);
+    return false;
+  }
+  return true;
+}
+
+static int
+store_decode_plgd_time(const oc_rep_t *rep, size_t device, void *data)
+{
+  (void)device;
+  plgd_time_t *pt = (plgd_time_t *)data;
+  if (!plgd_time_decode(rep, pt)) {
+    OC_ERR("cannot load plgd time: cannot decode representation");
+    return -1;
+  }
+  return 0;
+}
+
+bool
+plgd_time_load(void)
+{
+  plgd_time_t pt = { 0 };
+  if (oc_storage_load_resource(PLGD_TIME_STORE_NAME, 0, store_decode_plgd_time,
+                               &pt) <= 0) {
+    OC_ERR("failed to load plgd time from storage");
+    return false;
+  }
+
+  OC_DBG("plgd time loaded from storage");
+  return dev_time_set_time(pt.store.last_synced_time, false, false) == 0;
+}
+
+static void
+plgd_time_resource_post(oc_request_t *request, oc_interface_mask_t iface_mask,
+                        void *data)
+{
+  (void)iface_mask;
+  (void)data;
+  plgd_time_t pt = { 0 };
+  if (!plgd_time_decode(request->request_payload, &pt)) {
+    OC_ERR("cannot decode data for plgd time resource");
+    oc_send_response(request, OC_STATUS_BAD_REQUEST);
+    return;
+  }
+
+  if (dev_time_set_time(pt.store.last_synced_time, false, false) != 0) {
+    oc_send_response(request, OC_STATUS_INTERNAL_SERVER_ERROR);
+    return;
+  }
+
+  if (plgd_time_encode(g_oc_plgd_time, OC_IF_RW, false) != 0) {
+    OC_ERR("cannot encode plgd time resource");
+    oc_send_response(request, OC_STATUS_INTERNAL_SERVER_ERROR);
+    return;
+  }
+
+  oc_send_response(request, OC_STATUS_CHANGED);
+  plgd_time_dump();
+#if defined(OC_SERVER)
+  oc_notify_observers_delayed(oc_core_get_resource_by_index(PLGD_TIME, 0), 0);
+#endif /* OC_SERVER */
```

Comment:
This is not needed for POST because `OC_STATUS_CHANGED` automatically trigger notification